x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/ResultsIndexUpgradeService.java

```diff
@@ -0,0 +1,516 @@
+        UpdateSettingsRequest request = new UpdateSettingsRequest(indices.toArray(new String[0]));
+        request.settings(settings);
+        executeAsyncWithOrigin(client.threadPool().getThreadContext(),
+            ML_ORIGIN,
+            request,
+            listener,
+            client.admin().indices()::updateSettings);
+    }
+
+    private void removeReadOnlyBlock(Client client, List<String> indices,
+                                     ActionListener<AcknowledgedResponse> listener) {
+        Settings settings = Settings.builder().put(IndexMetaData.INDEX_READ_ONLY_SETTING.getKey(), false).build();
+        UpdateSettingsRequest request = new UpdateSettingsRequest(indices.toArray(new String[0]));
+        request.settings(settings);
+        executeAsyncWithOrigin(client.threadPool().getThreadContext(),
+            ML_ORIGIN,
+            request,
+            listener,
+            client.admin().indices()::updateSettings);
+    }
+
+    private void reindexOldReadIndicesToNewIndices(Client client,
+                                                   Map<String, String> reindexIndices,
+                                                   ResultsIndexUpgradeAction.Request request,
+                                                   ActionListener<Boolean> listener) {
+        TypedChainTaskExecutor<BulkByScrollResponse> chainTaskExecutor =
+            new TypedChainTaskExecutor<>(
+                client.threadPool().executor(executor),
+                (createIndexResponse) -> { // If there are errors in the reindex, we should stop
+                    Tuple<RestStatus, Throwable> status = getStatusAndCause(createIndexResponse);
+                    if (status.v1().equals(RestStatus.OK)) {
+                        return true;
+                    } else {
+                        // Throw an unchecked exception to keep from executing the rest of the reindexing requests
+                        //   and return the failure to the end user.
+                        throw new ElasticsearchException(status.v2());
+                    }
+                },
+                (exception -> true)); // Short circuit and call onFailure for any exception
+
+        List<String> newIndices = new ArrayList<>(reindexIndices.size());
+        reindexIndices.forEach((oldIndex, newIndex) -> {
+            ReindexRequest reindexRequest = new ReindexRequest();
+            reindexRequest.setSourceBatchSize(request.getReindexBatchSize());
+            reindexRequest.setSourceIndices(oldIndex);
+            reindexRequest.setDestIndex(newIndex);
+            // Don't worry if these indices already exist, we validated index.format earlier
+            reindexRequest.setConflicts("proceed");
+            // If the document exists already in the new index, don't want to update or overwrite as we are pulling from "old data"
+            reindexRequest.setDestOpType("create");
```

Comment:
I figured that we would not want to overwrite data from the "old" indices into the "new" ones, consequently we only worry about creates.